### PR TITLE
[WIP] Support chunks in HTTP2 Fixes #237

### LINKED ISSUE #237: Support chunks in HTTP2
Patrick at WebPageTest has added a super useful feature for websites that use HTTP2. He now includes when chunks of data are sent from the server in the HAR. Instead of just having those long blocks (start/stop) you can now see when chunks of data are sent (the chunks have darker color)

screen shot 2018-09-22 at 21 01 01

The response has extra data:

```
 "_chunks": [
 {
   "bytes": 385,
   "ts": 1361.457
},
{
   "bytes": 1300,
   "ts": 1367.293
},
```
Here's one full example:
https://gist.github.com/soulgalore/2a45fb2ecce8d8e039fd587ef7cb437c

I think this would be cool to support.

### PR DESCRIPTION
**Pull request by radum**
This the work I have done for fix #237. Is not yet finished.

### CODE REVIEW

**Comment by radum** [comment]
@soulgalore I guess we can continue our conversation from #237 here to this over the finish line 😁 

In theory everything works as it is, you can see the chunks rendering, but needs to be tested with some edge cases I guess.

Also the har entry interface is missing the chunks key so that will fail TS if we use it. We either have to copy and keep our own version of we raise it with them and hope it gets merged.

**Comment by soulgalore** [comment]
Great thank you @radum ! I have a look later this weekend.

File: src/ts/transformers/har.ts [hunk 1]
```diff
@@ -236,12 +236,15 @@ const getUserTimings = (currPage: Page, options: ChartOptions) => {
 /**
  * Create `WaterfallEntry`s to represent the sub-timings of a request
  * ("blocked", "dns", "connect", "send", "wait", "receive")
+ * TODO: `harEntry: Entry` Entry interface doesn't have the chunks property and this fails TS.
+ * I see no benefit in typechecking harEntry IMHO. But to fix this we need to import that interface and update it.
```
Comment (soulgalore) [inline src/ts/transformers/har.ts]: I totally agree with that. The thing is that the HAR specification says that you can add your own field as long as it starts with an underscore (check **Custom fields** in http://www.softwareishard.com/blog/har-12-spec/). @micmro would it be possible to fix that?

Comment (micmro) [reply src/ts/transformers/har.ts]: I'll have a look.

Comment (micmro) [reply src/ts/transformers/har.ts]: This would be a nice use-case for [Template Literal Types](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html), but this would force consuming project to use TypeScript version >= 4.1.

The problem of not not typing this is that a user would never know what properties they can rely on being set in Entry and which are optional. As for not non-standard fields, you could still add a [Type-Predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) (previously known as Type-Guard) to type-safely assert they field accurately exists:

E.g.:
```ts
type ChunkData = { bytes: number; ts: number };
const isEntityWithChunk = (harEntry: Entry): harEntry is Entry & {_chunk: ChunkData[]} => {
  return '_chunk' in harEntry
}

const buildDetailTimingBlocks = (startRelative: number, harEntry: Entry): WaterfallEntryTiming[] => {
  if(isEntityWithChunk(harEntry)){

    const chunks = harEntry._chunk
    ...
  }
  ...
}
```

I'll setup a PR to `@types/har-format` to add `_chunks` as non-standard type for now.

Comment (micmro) [reply src/ts/transformers/har.ts]: PR: for `@types/har-format` https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56428

Comment (micmro) [reply src/ts/transformers/har.ts]: Update is merged, `_chunk` is available with `1.2.8` now.

Comment (soulgalore) [reply src/ts/transformers/har.ts]: Great thank you @micmro :) I haven't had time to look at the PR (implemented the support in Chroma HAR though), let me do it this week.

**Comment by soulgalore** [comment]
@radum this looks good for me, was there something special you where thinking that needed to be fixed? I think the 1px instead of the "correct" width is fine, it will still help Ito show when content is sent?

**Comment by radum** [comment]
Apart from that I can't think of anything else @soulgalore. I just wanted to test it using lots of examples to make sure it works fine.

As for the 1px issue, yeah for sure it is better than nothing and I think still valuable. Not sure what will be the best way to bring the correct width in though. Although we can incrementally bring that in at a later date.

**Comment by soulgalore** [comment]
Works fine for me too, @micmro is it ok for you so we can merge/release it?

**Comment by micmro** [comment]
@radum would you mind updating `@types/har-format` to `1.2.8` to fix the `_chunk` type?

**Comment by radum** [comment]
@micmro Done. Not sure if you want to change the colors to look a bit better when the chunks are present. I think the contrast should be reversed. The light shade should be the bar and the dark one the chunks.

**Review by micmro — APPROVED** [review]

**Comment by micmro** [comment]
Thanks a lot @radum! I'll merge it and play around with the colours a bit and cut a new release after.

**Comment by micmro** [comment]
Released with `2.11.0`
